New module: GitHub Enterprise Organizations API [source_control/github_organization]

peter-murray opened the pull request:
Adding a module to provide the ability to interact with GitHub Enterp…rise organizations

##### SUMMARY
Creating a module that can create and rename GitHub Enterprise Organizations.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
github_organization

##### ANSIBLE VERSION
```
2.5.0
```


##### ADDITIONAL INFORMATION
Creating a module to interact with the GitHub Enterprise organisations API https://developer.github.com/v3/enterprise-admin/orgs/ to provide the ability to add and rename repositories in a GitHub Enterprise instance.

peter-murray committed:
Adding a module to provide the ability to interact with GitHub Enterprise organizations

peter-murray committed:
pep8 changes to pass test and clean up of module imports

peter-murray committed:
Correcting failing sanity checks

peter-murray committed:
Changing imports for failing tests in shippable

peter-murray commented:
What needs to be done to get this merged in? No one in the community has followed up on it, and I am willing to support it directly myself from the community perspective.

peter-murray commented:
bot_status

thaumos commented:
@peter-murray, please attend one of our Community IRC meetings to discuss this further. 

https://github.com/ansible/community/tree/master/meetings#schedule

peter-murray commented:
@thaumos Tried to attend today, but there did not appear to be a community meeting, the networking one ran over and I waited for another 40 minutes after that and nothing started.

I can see that the code has been frozen now for 2.4 release. When or what do I need to do to get this integrated, it has been open for so long now and I either want to get it integrated, or I will just abandon it altogether and go back to writing my own modules for my own use and those of the companies that I work for.

s-hertel commented on lib/ansible/modules/source_control/github_organization.py:
```diff
@@ -0,0 +1,308 @@
+#
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
```
New copyright: http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#copyright

s-hertel commented on lib/ansible/modules/source_control/github_organization.py:
```diff
@@ -0,0 +1,308 @@
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+ANSIBLE_METADATA = {'metadata_version': '1.0',
```
'1.1'

s-hertel commented on lib/ansible/modules/source_control/github_organization.py:
```diff
@@ -0,0 +1,308 @@
+short_description: Create/Rename a GitHub organization.
+description:
+    - Create or Rename a GitHub organization.
+version_added: "2.4"
```
"2.5"

s-hertel commented on lib/ansible/modules/source_control/github_organization.py:
```diff
@@ -0,0 +1,308 @@
+      -  Whether to create or rename an organization
+    choices: ['present', 'rename']
+    default: 'present'
+    required: false
```
You can remove this line. You don't need to document required: false as it is the default. Same for any following this.

s-hertel commented on lib/ansible/modules/source_control/github_organization.py:
```diff
@@ -0,0 +1,308 @@
+    description:
+      - Whether or not to validate the certificate returned from the GitHub Enterprise
+        instance. This is useful when using a self signed certificate.
+    default: True
```
May want to document that this is a `type: bool`

s-hertel commented on lib/ansible/modules/source_control/github_organization.py:
```diff
@@ -0,0 +1,308 @@
+
+    if state == 'present':
+        result = create_organization(module, auth_header)
+    elif state == 'rename':
```
Does having a 'rename' option for state make sense? It seems like you could just check if the organization exists and if it does and rename_to is provided, then rename it. Not requesting a change, just wondering.

peter-murray replied:
I chose to make rename a state due to the slightly odd behaviour of a rename action. See comment above about why I made rename a state.

I am open to suggestions, but I would probably see `state: rename` being used to clean up your GHE organisations as a one time event, rather than something that would be run continuously. A rename action requires the source organization to exist, and the rename to organization to not exist. Once a rename has taken place, the same action cannot be completed, as the source would not exist and target to rename to does.

I thought long and hard on this, but could not see a better way, as this is not really an idempotent action, but an important one to to fix up mistakes or errors.

bcoca replied:
i would require one of id/name for normal operation and require ID if you want to change name, which now can be a 'check if i need to' making it safe to rerun the same task.

s-hertel commented on lib/ansible/modules/source_control/github_organization.py:
```diff
@@ -0,0 +1,308 @@
+        headers = {'Authorization': auth_header}
+        data = {
+            "login": organization_name,
+            # "profile_name": organization_name,
```
Since having a display name is optional, you can update data if organization_name is provided.

Actually - since you've made name a required module option, you can just uncomment this line, right?

peter-murray replied:
The profile name is more like a description once the organisation is created, it is a fairly badly named parameter in the API. This corresponds to a "Organization display name" within the settings for an organization.

Most users (myself very much included) would likely prefer to leave this value unset, as not setting this value just uses the organization name within the UI.

I left it here just in case it may be desired in the future to add it back in via a parameter.

s-hertel commented on lib/ansible/modules/source_control/github_organization.py:
```diff
@@ -0,0 +1,308 @@
+            result.update({'changed': True})
+            _merge_organization_result(result, org)
+        else:
+            failure = module.from_json(info['body'])
```
Will info['body'] be json? Might be a string/HTML.

peter-murray replied:
The GitHub API should only return JSON here, it is documented as such. I can go more defensive with try/catch but it should only be returning JSON here.

sivel replied:
Someone could have a proxy in front, and if it intercepted a non-200 status code could replace the body with HTML or otherwise.

peter-murray replied:
Good point!

s-hertel commented on lib/ansible/modules/source_control/github_organization.py:
```diff
@@ -0,0 +1,308 @@
+
+
+def _get_auth_header(user, token):
+    auth = base64.encodestring('%s:%s' % (user, token)).replace('\n', '')
```
Can you test this on python3 please? I think python3 needs bytes.

s-hertel commented on lib/ansible/modules/source_control/github_organization.py:
```diff
@@ -0,0 +1,308 @@
+                failure = module.from_json(info['body'])
+                module.fail_json(msg='Failed to rename organization: {0}'.format(failure['message']))
+        else:
+            module.fail_json(msg="An organization with the target name already exists, '{0}'".format(new_name))
```
Rather than failing here, should this just return the target org info with changed: False for idempotent behavior?

peter-murray replied:
That is for a rename, if you request to rename something, and the target name already exists, the renaming cannot be completed.

To me that is a failure, you asked for `x` to become `y`, but this cannot be done as `y` is already taken.

This is why I made rename a state.

s-hertel commented on lib/ansible/modules/source_control/github_organization.py:
```diff
@@ -0,0 +1,308 @@
+  state:
+    description:
+      -  Whether to create or rename an organization
+    choices: ['present', 'rename']
```
Is it possible to remove an organization?

peter-murray replied:
No, the API does not support removal. I have raised a feature request on this with GitHub, if enough people request the feature, it will be implemented, but there is no API to cover this in 2.11, only create and rename as per https://developer.github.com/v3/enterprise-admin/orgs/

s-hertel commented on lib/ansible/modules/source_control/github_organization.py:
```diff
@@ -0,0 +1,308 @@
+        result = create_organization(module, auth_header)
+    elif state == 'rename':
+        result = rename_organization(module, auth_header)
+    else:
```
You can remove the else. AnsibleModule will handle this since state has a list of allowed choices.

s-hertel commented on lib/ansible/modules/source_control/github_organization.py:
```diff
@@ -0,0 +1,308 @@
+    existing_org = _get_organization(module, base_url, auth_header, organization_name)
+    target_org = _get_organization(module, base_url, auth_header, new_name)
+
+    if existing_org is not None:
```
If the organization being renamed doesn't exist yet, that should probably throw an error or a warning or something.

peter-murray committed:
Changes as requested from code review on Pull Request #26071

peter-murray committed:
Dealing with a potential non-JSON return type on errors

peter-murray committed:
Resolving failing boilerplate sanity test failure

peter-murray committed:
Fixing import failure test

peter-murray commented:
@s-hertel I have addressed the points that you raised either by code changes or providing extra information to your questions in the code review, and now CI validation checks had passed.

s-hertel commented:
@peter-murray Should the module fail if the user is trying to rename an org that doesn't exist? (the comment on line 253 - https://github.com/ansible/ansible/pull/26071#discussion_r141404563). Besides that, this looks good to me! Sorry this wasn't reviewed sooner.

peter-murray committed:
Failing on a rename when the organization to rename is not found

peter-murray commented:
@s-hertel That is where the `rename` state starts to bite me as it is a bit of a hack... Does a rename that does not change due to not finding the original organization count as a failure, or just that it did not change?

I think it should fail as you highlighted, based on the slightly edge case not quite as idempotent as I would like for a state. The problem would be if you you decided to rename a lot of organizations, some pass, it then fails and you re-run that playbook, now you end up with failures due to some of them succeeding... As you can tell I was clearly on the fence with how that would work in practice.

I truly appreciate you taking the time to review this.

bcoca commented on lib/ansible/modules/source_control/github_organization.py:
```diff
@@ -0,0 +1,301 @@
+    description:
+      - The owner of the organization that is being created.
+        The owner is required when I(state=present)
+  rename_to:
```
no need for state=rename ... if rename_to is not None ... you want to rename.

bcoca replied:
I would go one step further and use id as unique identifier and change name if it is different, avoiding rename_to/state=rename altogether

peter-murray replied:
Thanks, for the feedback, but when I went to implement it in that way, it does not hold together as nicely as it might first appear.

So we now have an `id` parameter that is the organization to create, you would typically not provide a `name` for creation, but you could do so.

If `id` and `name` are both provided, then we check if they are the same value.
If they are a the same, it is effectively a *create*, and will create the organization if it does not exist, otherwise doing nothing.

If they differ, then we have a *rename* and the `id` is used for the source organization to be renamed to `name` value. This is where we have the issue, as now effectively the old `id` no longer exists (as it was changed to `name`) if you re-run the playbook/action.
I fail the *rename* attempt if the source organization cannot be found, which I think is a requirement of that type of action, as *changed: False* does not feel right if you cannot find the thing to *rename*, as it is misleading (at least from my perspective).

I am on the fence about possibly disabling the *rename* feature, but I know that I would have used it in the past (which was why I added it).

peter-murray replied:
@bcoca Are you happy if I only refactor to the first suggestion due to what I commented on above?

maxamillion replied:
@bcoca friendly re-ping

gundalow commented:
Hi, we now have a dedicated `source_control/github/` directory, could you please rebase and move the module there? Thanks in advance.
